agents/go-agents/bootstrapper/booter/installations.go

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright (c) 2012-2017 Codenvy, S.A.
+// All rights reserved. This program and the accompanying materials
+// are made available under the terms of the Eclipse Public License v1.0
+// which accompanies this distribution, and is available at
+// http://www.eclipse.org/legal/epl-v10.html
+//
+// Contributors:
+//   Codenvy, S.A. - initial API and implementation
+//
+
+package booter
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/eclipse/che/agents/go-agents/core/process"
+)
+
+// Process of getting certain installer software/servers installed.
+type installation interface {
+
+	// Check installation preconditions e.g. ports are free.
+	// If error is returned installation is not executed.
+	preCheck() error
+
+	// Executes the installation in implementation specific way.
+	execute() error
+}
+
+// Script installation executes script defined by installer.
+// Installation is considered successful only if script exit code is 0.
+type scriptInst struct {
+	installer Installer
+	timeout   time.Duration
+}
+
+func (sci *scriptInst) preCheck() error { return nil }
+
+func (sci *scriptInst) execute() error {
+	_, diedC, err := executeScript(sci.installer)
+	if err != nil {
+		fmt.Errorf("Scrip execution failed. Error: %s", err)
+	}
+	select {
+	case died := <-diedC:
+		if died.ExitCode != 0 {
+			return fmt.Errorf(
+				"Exit code for '%s' installation is '%d' while expected to be 0",
+				sci.installer.ID,
+				died.ExitCode,
+			)
+		}
+	case <-time.After(sci.timeout):
+		return fmt.Errorf("Timeout reached before installation of '%s' completed", sci.installer.ID)
+	}
+	return nil
+}
+
+// Server installation executes script defined by installer and
+// wait all the defined servers to become available.
+type serverInst struct {
+	installer Installer
+	period    time.Duration
+	timeout   time.Duration
+}
+
+func (svi *serverInst) preCheck() error {
+	for _, server := range svi.installer.Servers {
+		if tryConn(server) == nil {
+			return fmt.Errorf("server address 'localhost:%s' already in use", server.Port)
```

Comment:
I suppose that already running server is not a problem for an agent start. For example, [ubuntu_jdk8 image](https://hub.docker.com/r/eclipse/ubuntu_jdk8/~/dockerfile/) start SSHD as the main command of the container. So with this check SSH agent will fail on that image.